sdk/core/System.ClientModel/src/Pipeline/ClientLoggingPolicy.cs

```diff
@@ -0,0 +1,455 @@
+using System.Threading.Tasks;
+
+namespace System.ClientModel.Primitives;
+
+/// <summary>
+/// TODO.
+/// </summary>
+public class ClientLoggingPolicy : PipelinePolicy
+{
+    private const double RequestTooLongTime = 3.0; // sec
+    private const int DefaultLoggedContentSizeLimit = 4 * 1024;
+    private static ClientModelEventSource? s_eventSource;
+    private const string DefaultEventSourceName = "System.ClientModel";
+
+    private readonly bool _logContent;
+    private readonly int _maxLength;
+    private readonly string? _assemblyName;
+    private readonly string? _clientRequestIdHeaderName;
+    private readonly bool _isLoggingEnabled;
+
+    internal static ClientModelEventSource Singleton { get => s_eventSource ??= ClientModelEventSource.Create(DefaultEventSourceName, Array.Empty<string>()); } //TODO add traits
+
+    /// <summary>
+    /// TODO.
+    /// </summary>
+    /// <param name="eventSourceName"></param>
+    /// <param name="eventSourceTraits"></param>
+    /// <param name="options"></param>
+    protected ClientLoggingPolicy(string? eventSourceName, string[]? eventSourceTraits = default, DiagnosticsOptions? options = default) : this(options)
+    {
+        if (s_eventSource != null && !string.IsNullOrEmpty(eventSourceName) && eventSourceName != s_eventSource.Name)
+        {
+            throw new ArgumentException("Cannot use multiple event source names in the same application.");
+        }
+
+        if (s_eventSource == null)
+        {
+            s_eventSource = ClientModelEventSource.Create(eventSourceName ?? DefaultEventSourceName, eventSourceTraits);
+        }
+    }
+
+    /// <summary>
+    /// TODO.
+    /// </summary>
+    /// <param name="options"></param>
+    public ClientLoggingPolicy(DiagnosticsOptions? options = default)
+    {
+        _logContent = options?.IsLoggingContentEnabled ?? false;
+        _maxLength = options?.LoggedContentSizeLimit ?? DefaultLoggedContentSizeLimit;
+        _assemblyName = options?.LoggedClientAssemblyName;
+        _clientRequestIdHeaderName = options?.RequestIdHeaderName;
+        _isLoggingEnabled = options?.IsLoggingEnabled ?? true;
+    }
+
+    /// <summary>
+    /// TODO.
+    /// </summary>
+    internal PipelineMessageSanitizer? Sanitizer { get; set; }
+
+    /// <inheritdoc/>
+    public override void Process(PipelineMessage message, IReadOnlyList<PipelinePolicy> pipeline, int currentIndex) =>
+        ProcessSyncOrAsync(message, pipeline, currentIndex, async: false).EnsureCompleted();
+
+    /// <inheritdoc/>
+    public override async ValueTask ProcessAsync(PipelineMessage message, IReadOnlyList<PipelinePolicy> pipeline, int currentIndex) =>
+        await ProcessSyncOrAsync(message, pipeline, currentIndex, async: true).ConfigureAwait(false);
+
+    private async ValueTask ProcessSyncOrAsync(PipelineMessage message, IReadOnlyList<PipelinePolicy> pipeline, int currentIndex, bool async)
+    {
+        if (!_isLoggingEnabled || !Singleton.IsEnabled())
```

Comment:
You know ... if IsLoggingEnabled goes on LoggingOptions/DiagnosticsOptions, maybe we want to do what Azure.Core does and not add the logging policy into the pipeline if it's false?  What do you think?